atomix/cluster/src/main/java/io/atomix/raft/roles/FollowerRole.java

```diff
@@ -188,61 +186,50 @@ private void sendPollRequests() {
   @Override
   public CompletableFuture<ConfigureResponse> onConfigure(final ConfigureRequest request) {
     final CompletableFuture<ConfigureResponse> future = super.onConfigure(request);
-    resetHeartbeatTimeoutFromDifferentThread();
+    if (isRequestFromCurrentLeader(request.term(), request.leader())) {
+      resetHeartbeatTimeout();
+    }
     return future;
   }
 
   @Override
   public CompletableFuture<AppendResponse> onAppend(final AppendRequest request) {
-    // Reset the heartbeat timeout.
-    resetHeartbeatTimeoutFromDifferentThread();
-
-    return super.onAppend(request);
+    final CompletableFuture<AppendResponse> future = super.onAppend(request);
+    if (isRequestFromCurrentLeader(request.term(), request.leader())) {
```

Comment:
Should we reset it after the append is complete? Shouldn't we acknowledge it immediately?